Tests/xcodeprojMeasurementsTests/MeasurementsTests.swift

```diff
@@ -0,0 +1,97 @@
+import Foundation
+import PathKit
+import SwiftShell
+@testable import xcodeproj
+import XCTest
+
+final class MeasurementsTests: XCTestCase {
+    
+    let tempDirectory = Path.temporary
+    var projectPath: Path?
+    var sourceRoot: Path?
+    
+    override func setUp() {
+        super.setUp()
+        try? tempDirectory.delete()
+        self.projectPath = tempDirectory + Path("Project")
+        self.sourceRoot = tempDirectory + Path("Sources")
+        let projectPath = self.projectPath!
+        let sourceRoot = self.sourceRoot!
+        let xcodeproj = try! generateXcodeProj(path: projectPath, sourceRoot: sourceRoot)
+        validate(xcodeproj: xcodeproj)
+    }
+    
+    override func tearDown() {
+        super.tearDown()
+        try? tempDirectory.delete()
+    }
+    
+    func test_measureProjectRead() {
+        let duration = try! measureDuration {
+            _ = try XcodeProj(path: self.projectPath!)
+        }
+        XCTAssertLessThan(duration, 3)
+    }
+    
+    func test_measureProjectWrite() {
+        let xcodeproj = try! XcodeProj(path: self.projectPath!)
+        let duration = try! measureDuration {
+            try xcodeproj.write(path: self.projectPath!)
+        }
+        XCTAssertLessThan(duration, 5)
+    }
+    
+    func test_measureObtainFilesFullPath() {
+        let xcodeproj = try! XcodeProj(path: self.projectPath!)
+        var pathList = [Path]()
+        var count = 0
+        let duration = try! measureDuration {
+            for project in xcodeproj.pbxproj.projects {
+                for target in project.targets {
+                    for buildPhase in target.buildPhases {
+                        let files = buildPhase.files
+                        for file in files {
+                            let fullPath = try file.file!.fullPath(
+                                sourceRoot: sourceRoot!
+                            )
+                            pathList.append(fullPath!)
+                            count += 1
+                        }
+                    }
+                }
+            }
+        }
+        XCTAssertLessThan(duration, 0.1)
+    }
+    
+    func test_measureProjGeneration() {
```

Comment:
@pepibumur And one more problem. Generation of the project is very long. (more than 10 minutes) 
The reason is another performance issue.
`PBXGroup func addFile(...)` This method takes a very long time. the more files the longer it takes.
This is due to this logic:
```
if let existingFileReference = try projectObjects.fileReferences.first(where: {
            try filePath == $0.value.fullPath(sourceRoot: sourceRoot)
 })
```
If I remove this logic, the generation will take less than half a minute.
A very long time to sort through all the files. Maybe there are some alternatives? Alternatively, we can create an utility dictionary where the key is the full path. What do you think about this option?

But perhaps this logic is not needed at all.
I do not fully understand why we change the path of the existing fileReference and why create a new fileReference if there was already an old fileReference.

Reply:
The `addFile` method is intended to be used exceptionally when someone needs to add a single file to an existing group. It makes the addition of files convenient but at the cost of a not so good performance. Projects that want to have a high-performance project generation should not use those convenient methods and modify the objects directly.

I suggest changing the approach and adding a comment that states that the method shouldn't be used for adding a large list of files.